### ISSUE TITLE
Benchmark v0.0 - Microsoft Flood and Clouds Segmentation

### ISSUE THREAD

**Issue by lillythomas** [Clay-foundation/model#117]
We've completed our first benchmarking exercise for Clay v0 using the Microsoft Flood detection dataset (see https://github.com/Clay-foundation/model/issues/83 for details).

The exercise entailed a comparison of a model "trained from scratch" vs. a model finetuned from Clay. They both are designed to segment input into binary maps of flood and background. The data module ingests datacubes  like those used to train clay, with the addition of water masks and cloud/cloud shadow masks, applies cloud/cloud shadow masking and adds the water masks (labels) into the batch dictionary.

The from scratch model is a simple encoder / decoder architecture with He initialization, dropout regularization, learning rate scheduling (`lr=1e-3` to start) and early stopping. The loss function is binary cross entropy.

The finetuned model uses Clay as the encoder and uses a simple multi layer convolutional decoder. It also implements learning rate scheduling, early stopping and minimizes a binary cross entropy loss function.

# Results
The model from scratch converged (based on validation loss) in **23 epochs**, whereas the model finetuned from Clay converged in **19 epochs**.

#### From scratch
Train loss
<img width="747" alt="Screen Shot 2024-01-12 at 9 15 41 PM" src="https://github.com/Clay-foundation/model/assets/18336080/b3b4dfef-0b8a-492e-b58c-a7e5387e1375">

Val loss
<img width="749" alt="Screen Shot 2024-01-12 at 9 15 49 PM" src="https://github.com/Clay-foundation/model/assets/18336080/7f873dd2-e749-4486-a140-1cd2d62b8f01">

#### Finetuned
Train loss
<img width="748" alt="Screen Shot 2024-01-12 at 9 13 18 PM" src="https://github.com/Clay-foundation/model/assets/18336080/958974e6-fe4c-47a8-9c75-8e365bdd9d86">

Val loss
<img width="746" alt="Screen Shot 2024-01-12 at 9 13 32 PM" src="https://github.com/Clay-foundation/model/assets/18336080/bd279767-ac40-4416-b78e-84ae6073c79e">

## Evaluation metric
Out evaluation entailed calculating **mean IoU** on the validation dataset.

#### From scratch
mIoU =  `0.27`

#### Finetuned
mIoU = `0.31` ❗ 

## Visual comparison

Blue band on left, water mask label in middle, prediction on right.

### From scratch

<img width="750" alt="Screen Shot 2024-01-12 at 8 00 17 PM" src="https://github.com/Clay-foundation/model/assets/18336080/1a459c8e-7e87-4d6f-b878-1760aeda5e6b">
<img width="750" alt="Screen Shot 2024-01-12 at 7 58 39 PM" src="https://github.com/Clay-foundation/model/assets/18336080/b7abccbf-5fbd-492f-b87d-be345a62412d">
<img width="750" alt="Screen Shot 2024-01-12 at 7 57 30 PM" src="https://github.com/Clay-foundation/model/assets/18336080/f2bf5c11-8422-4000-95e4-9df53b1bc999">
<img width="750" alt="Screen Shot 2024-01-12 at 9 18 16 PM" src="https://github.com/Clay-foundation/model/assets/18336080/975c2e09-e3ba-45ea-a271-6b8ce9acddc4">
<img width="750" alt="Screen Shot 2024-01-12 at 9 21 41 PM" src="https://github.com/Clay-foundation/model/assets/18336080/84be8283-78a7-4cf5-929b-97f710670a04">

### Finetuned

<img width="750" alt="Screen Shot 2024-01-12 at 7 59 38 PM" src="https://github.com/Clay-foundation/model/assets/18336080/3d20e31d-6b4f-4bfc-83a6-f31fc4be6fd0">
<img width="750" alt="Screen Shot 2024-01-12 at 7 55 16 PM" src="https://github.com/Clay-foundation/model/assets/18336080/0815e396-58ae-4585-b826-9944ddedb83e">
<img width="750" alt="Screen Shot 2024-01-12 at 7 56 09 PM" src="https://github.com/Clay-foundation/model/assets/18336080/c71adec7-08d1-4f18-a8a4-f9c8c2cd83e0">
<img width="750" alt="Screen Shot 2024-01-12 at 9 17 04 PM" src="https://github.com/Clay-foundation/model/assets/18336080/1f28e038-87b9-45c6-bfa9-96635eaa54c2">
<img width="750" alt="Screen Shot 2024-01-12 at 9 21 26 PM" src="https://github.com/Clay-foundation/model/assets/18336080/5a0f28be-3ee9-41b9-8e2d-72d48058608d">

Notebook: https://github.com/Clay-foundation/model/blob/benchmark_segmentation_task/notebooks/flood_benchmark_segmentation.ipynb

**Comment by lillythomas** [comment]
I've done a few more experiments to rigor-test things. The main changes involved shuffling the samples before partitioning using a seed for consistency across model variants and reproducibility. The prior metrics in the first comment did not involve shuffled data pre-partitioning. These experiments demonstrate how much the results vary based on different data distributions. 

### Varying data distributions

| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 1   | From scratch   | 46   | 0.57   | 42   |
| 1   | Finetuned   | 23   | 0.42   | 42   |
| 2   | From scratch   | 36   | 0.44   | 33   |
| 2   | Finetuned   | 8   | 0.31   | 33   |
| 3   | From scratch   | 30   | 0.38   | 21   |
| 3   | Finetuned   | 11   | 0.33   | 21   |
| 4   | From scratch| 19   | 0.48   | 9   |
| 4   | Finetuned   | 26   | 0.38   | 9  |

In the above table, finetuned refers to _finetuned with a multi-layer convolutional decoder_.

### Using the clay decoder
An experiment in which the multi-layer convolutional decoder was replaced with the Clay decoder. 
| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 5   | From scratch   | 36   | 0.44   | 33   |
| 5   | Finetuned (conv decoder)   | 8   | 0.31   | 33   |
| 5   | Finetuned (clay decoder)   | 369| 0.33   | 33   |


### Controlling the number of epochs

An experiment where the number of epochs were standardized instead of determined by an early stopping callback.

For a set number of 10 epochs, we achieved:
| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 6  | From scratch   | 10   | 0.14  | 33   |
| 6   | Finetuned (conv decoder)   | 10   | 0.30   | 33   |
| 6   | Finetuned (clay decoder)   | 10   | 0.41   | 33   |

All in all, the Clay decoder performs better than the multi-layer convolutional decoder and it is likely that overall, the `clay-finetuned + clay decoder` performs better when controlling for training time than the `from scratch` variant.

**Comment by lillythomas** [comment]
### Controlling the number of epochs (part ii)

An experiment where the number of epochs were standardized instead of determined by an early stopping callback.

For a set number of 30 epochs, we achieved:
| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 7  | From scratch   | 30   | 0.55  | 42   |
| 7   | Finetuned (conv decoder)   | 30 | 0.52   | 42   |
| 7   | Finetuned (clay decoder)   | 30   | 0.53   | 42   |


#### Loss curves
**From scratch**
<img width="750" alt="Screen Shot 2024-01-21 at 10 33 59 PM" src="https://github.com/Clay-foundation/model/assets/18336080/c5dfb253-8d9c-4623-be0d-81c72a8a8dfb">
<img width="750" alt="Screen Shot 2024-01-21 at 10 34 21 PM" src="https://github.com/Clay-foundation/model/assets/18336080/e19bc5d6-910c-4594-ae9b-29be88bbbdf1">

**Finetuned (conv and clay decoder)** 
<img width="750" alt="Screen Shot 2024-01-21 at 10 16 23 PM" src="https://github.com/Clay-foundation/model/assets/18336080/8d3d5770-1406-4bbe-911d-3765462d6373">
<img width="750" alt="Screen Shot 2024-01-21 at 10 16 02 PM" src="https://github.com/Clay-foundation/model/assets/18336080/52ac64a8-aeb4-45e1-88bf-a040ff5a1f4c">
36080/671e5636-0ed5-476d-8a91-49783aa6a988">

#### Visual examples
**From scratch**
<img width="750" alt="Screen Shot 2024-01-21 at 11 02 50 PM" src="https://github.com/Clay-foundation/model/assets/18336080/a006e72b-b06c-465c-bc91-6f66f2eb21b8">
<img width="750" alt="Screen Shot 2024-01-21 at 11 17 05 PM" src="https://github.com/Clay-foundation/model/assets/18336080/9cc54686-828b-4eb1-88d9-de91e115d237">
<img width="750" alt="Screen Shot 2024-01-21 at 11 23 12 PM" src="https://github.com/Clay-foundation/model/assets/18336080/44f778e4-6d2a-48da-be1b-c94dd39865bf">
<img width="750" alt="Screen Shot 2024-01-21 at 11 23 54 PM" src="https://github.com/Clay-foundation/model/assets/18336080/948bc994-37ef-4bed-99f8-e0d0d5f9a834">
<img width="750" alt="Screen Shot 2024-01-21 at 11 28 14 PM" src="https://github.com/Clay-foundation/model/assets/18336080/97f5b9e3-0d69-4b85-85d5-382c25106824">



**Finetuned (conv decoder)** 
<img width="750" alt="Screen Shot 2024-01-21 at 10 54 45 PM" src="https://github.com/Clay-foundation/model/assets/18336080/28302e2d-24fe-4e41-85ae-5a31f67b775c">
<img width="750" alt="Screen Shot 2024-01-21 at 11 17 37 PM" src="https://github.com/Clay-foundation/model/assets/18336080/aa37d5dd-f5d7-4fe2-a89f-7dcc1c68a188">
<img width="750" alt="Screen Shot 2024-01-21 at 11 19 38 PM" src="https://github.com/Clay-foundation/model/assets/18336080/d19ec7d0-89da-402c-9ed6-1050e6410308">
<img width="750" alt="Screen Shot 2024-01-21 at 11 24 28 PM" src="https://github.com/Clay-foundation/model/assets/18336080/aabe2b3c-0a09-4b8a-998a-ee7fcaabb7d9">
<img width="750" alt="Screen Shot 2024-01-21 at 11 28 45 PM" src="https://github.com/Clay-foundation/model/assets/18336080/92265778-a011-45f3-a0f4-55454bf93c00">


**Finetuned (clay decoder)** 
<img width="750" alt="Screen Shot 2024-01-21 at 10 54 25 PM" src="https://github.com/Clay-foundation/model/assets/18336080/211b68c5-5efa-4f07-9cf3-43a2639dc959">
<img width="750" alt="Screen Shot 2024-01-21 at 11 17 59 PM" src="https://github.com/Clay-foundation/model/assets/18336080/27908029-e77f-4c91-bc12-f83f870605c7">
<img width="750" alt="Screen Shot 2024-01-21 at 11 19 07 PM" src="https://github.com/Clay-foundation/model/assets/18336080/1447a3b3-e1b8-4d8e-856c-7cffb44e05ba">
<img width="750" alt="Screen Shot 2024-01-21 at 11 25 12 PM" src="https://github.com/Clay-foundation/model/assets/18336080/c72c5a7b-3529-4c47-814b-a4960b8579ef">
<img width="750" alt="Screen Shot 2024-01-21 at 11 29 05 PM" src="https://github.com/Clay-foundation/model/assets/18336080/e57baddd-1825-4081-a605-3dc94300136d">

**Comment by lillythomas** [comment]
Same part ii epoch-controlled experiment with two more seeds

| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 8  | From scratch   | 30   | 0.41  | 21   |
| 8   | Finetuned (conv decoder)   | 30 | 0.40   | 21   |
| 8   | Finetuned (clay decoder)   | 30   | 0.44   | 21   |

#### Loss curves
**From scratch**
<img width="750" alt="Screen Shot 2024-01-22 at 10 19 44 AM" src="https://github.com/Clay-foundation/model/assets/18336080/119073d2-4744-44d2-a8c3-acbb5cdab63a">
<img width="750" alt="Screen Shot 2024-01-22 at 10 19 57 AM" src="https://github.com/Clay-foundation/model/assets/18336080/62ad7e60-6c07-4c61-a489-fefe22e51e73">
**Finetuned (conv and clay decoders)**
<img width="750" alt="Screen Shot 2024-01-22 at 10 21 51 AM" src="https://github.com/Clay-foundation/model/assets/18336080/f5df9a24-600e-49e1-b4a0-b41f86f57600">
<img width="750" alt="Screen Shot 2024-01-22 at 10 22 36 AM" src="https://github.com/Clay-foundation/model/assets/18336080/f47581c2-f71a-4bed-bc75-b85f787303c8">

| Experiment | Variant | Number of epochs | Mean IoU | Seed |
| ---------| ---------| ---------| ---------| ---------|
| 9  | From scratch   | 30   | 0.44  | 33   |
| 9   | Finetuned (conv decoder)   | 30 | 0.42   | 33   |
| 9   | Finetuned (clay decoder)   | 30   | 0.47   | 33   |

#### Loss curves
**From scratch**
<img width="750" alt="Screen Shot 2024-01-22 at 2 44 51 PM" src="https://github.com/Clay-foundation/model/assets/18336080/08935d8b-a204-47b7-a28c-26c56d127b45">
<img width="750" alt="Screen Shot 2024-01-22 at 2 45 09 PM" src="https://github.com/Clay-foundation/model/assets/18336080/e3c5f942-df4d-494f-a715-f4c5d46c9247">
**Finetuned (conv and clay decoders)**
<img width="750" alt="Screen Shot 2024-01-22 at 2 43 37 PM" src="https://github.com/Clay-foundation/model/assets/18336080/24cbc154-e9fb-472b-a3f1-abb0eabfcc2a">
<img width="750" alt="Screen Shot 2024-01-22 at 2 44 02 PM" src="https://github.com/Clay-foundation/model/assets/18336080/38c22cc3-bd77-4b58-a0d1-d09269ae2249">


## Parameter counts per model

 | Name        | Type                      | Params
| model       | FloodDetector_fromscratch | 713 K 
| dice_metric | Dice                      | 0     
| 713 K     Trainable params
| 0         Non-trainable params
| 713 K     Total params
| 2.853     Total estimated model params size (MB)

 | Name      | Type              | Params
| model     | FloodDetector_clay_convdecoder    | 98.3 M
| criterion | BCEWithLogitsLoss | 0     
| 3.0 M     Trainable params
| 95.3 M    Non-trainable params
| 98.3 M    Total params
| 393.032   Total estimated model params size (MB)

 | Name      | Type              | Params
| model     | FloodDetector_clay_decoder     | 127 M 
| criterion | BCEWithLogitsLoss | 0     
| 32.4 M    Trainable params
| 95.3 M    Non-trainable params
| 127 M     Total params
| 510.809   Total estimated model params size (MB)
